nfs/src/rpc_task.cpp

```diff
@@ -304,6 +501,294 @@ void rpc_task::run_lookup()
     } while (rpc_retry);
 }
 
+std::vector<bytes_chunk>
+copy_to_cache(struct nfs_client *const client,
+        fuse_ino_t ino, struct fuse_bufvec* bufv, off_t offset, size_t length, int &error)
+{
+    // Check the valid ino.
+    assert(ino != 0);
+
+    const char* buf = nullptr;
+
+    // Set flag whether buffer we need to read from fd or not.
+    int fd_set = bufv->buf[bufv->idx].flags & FUSE_BUF_IS_FD;
+
+    // Whether the fd allow seek, we can use pread.
+    int seek_set = bufv->buf[bufv->idx].flags & FUSE_BUF_FD_SEEK;
+
+    // Whether the fd allow retry or not, in case of less buffer read.
+    int retry_set = bufv->buf[bufv->idx].flags & FUSE_BUF_FD_RETRY;
+    
+    // If fd is set then no buffer available.
+    if (fd_set) {
+        // assert(bufv->buf[bufv->idx].mem == 0x0);
+    } else {
+        buf = (char *) bufv->buf[bufv->idx].mem  + bufv->off;
+    }
+
+    // Get file_inode structure from inode and chunk vec.
+    auto file_inode = client->get_nfs_inode_from_ino(ino);
+    auto chunkvec = file_inode->filecache_handle->get(offset, length);
+    
+    // Result of pread/read.
+    int res = 0;
+    size_t rem_len = length;
+
+    // offset in fd from where to read.
+    off_t src_off = bufv->buf[bufv->idx].pos + bufv->off;
+    
+    for (auto &chunk : chunkvec)
+    {
+        auto membuf = chunk.get_membuf();
+
+        // Lock the membuf to do the operation.
+        membuf->set_locked();
+
+        // Chunk is owned by us or uptodate.
+        if (chunk.is_empty || membuf->is_uptodate())
+        {
+            // Read from the fd and copy to buffer.
+            if (fd_set) {
+                auto len = chunk.length;
+                auto dst_off = 0;
+
+                // Let's read in a loop.
+                while(len) {    
+                    if (seek_set) {
+                        res = pread(bufv->buf[bufv->idx].fd, chunk.get_buffer() + dst_off, len, src_off);
+                    } else {
```

Comment:
I think the fd mode is provided for splicing. Unless we use splice() it's not beneficial to use the fd mode. We can as well turn it off by not setting FUSE_CAP_SPLICE_READ. Did you try that?

Reply:
FUSE_CAP_SPLICE_READ worked, when I was implementing, tried setting FUSE_CAP_SPLICE_WRITE off which didn't work.